File: 0101-0200/0166-Fraction to Recurring Decimal/0166-Fraction to Recurring Decimal.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        result = []
        if numerator // denominator < 0:
            result.append('-')

        num = abs(numerator)
        den = abs(denominator)
        result.append(str(num // den))
        rem = num % den
        if rem == 0:
            return ''.join(result)
        result.append('.')
        table = {}
        while rem:
            table[rem] = len(result)
            rem *= 10
            result.append(str(rem // den))
            rem %= den
            if rem in table:
                result.insert(table[rem], '(')
                result.append(')')
                break
        return ''.join(result)

File: 0101-0200/0166-Fraction to Recurring Decimal/0166-Fraction to Recurring Decimal.cpp
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        string result = "";
        if ((float)numerator / denominator < 0) {
            result += '-';
        }
        
        long num = abs((long)numerator);
        long den = abs((long)denominator);
        
        result += to_string(num / den);
        long rem = num % den;
        if (rem == 0) {
            return result;
        }
        
        result += '.';
        unordered_map<long, int> table;        
        while(rem) {
            table[rem] = result.length();
            rem *= 10;
            result += to_string(rem / den);
            rem %= den;
            if (table.find(rem) != table.end()) {
                result.insert(table[rem], "(");
                result += ')';
                break;
            }
        }
        
        return result;        
    }
};
